src/assets/src/components/preferences.tsx

```diff
@@ -0,0 +1,168 @@
+            checked={notifyMeAttendee} 
+            onChange={() => setNotifyMeAttendee(!notifyMeAttendee)}
+            label="As an attendee, notify me via SMS when it becomes my turn." />
+    );
+    const notifyMeHostInput = (
+        <Form.Check 
+            type="checkbox"
+            id="notify-me-host"
+            disabled={!hasStartedEnteringPhoneNumber} 
+            checked={notifyMeHost}
+            onChange={() => setNotifyMeHost(!notifyMeHost)}
+            label="As a host, notify me via SMS when someone joins my empty queue." />
+    );
+
+    const validateAndSubmit = (e: React.SyntheticEvent) => {
+        e.preventDefault() // Prevent page reload
+        if (!changed) {
+            setValidationStatus(null);
+            return;
+        }
+        if (!hasStartedEnteringPhoneNumber) {
+            setValidationStatus([]);
+            props.onUpdateInfo('', notifyMeAttendee, notifyMeHost);
+            return; // Cleared form, so submit empty without validating
+        }
+        const validationErrors = validatePhoneNumber(phoneField, countryDialCode);
+        setValidationStatus(validationErrors);
+        if (!validationErrors.length)
+            props.onUpdateInfo(phoneField, notifyMeAttendee, notifyMeHost);
+    }
+
+    const alertBlock =
+        validationStatus === undefined // not yet validated
+            ? undefined
+        : validationStatus === null
+            ? <Alert variant='primary'>Your preferences were not changed.</Alert>
+        : validationStatus.length
+            ? <Alert variant='danger'>{validationStatus.map(e => <p>{e.message}</p>)}</Alert>
+        : props.errorOccurred
+            ? <Alert variant='danger'>An error occurred while trying to update your preferences; please try again later.</Alert>
+        : <Alert variant='success'>Your preferences were successfully updated.</Alert>
+
+    return (
+        <div>
+            <h1>View/Update Preferences</h1>
+            {alertBlock}
+            <Form onSubmit={validateAndSubmit}>
+                <p>Enter a phone number in order to opt in to SMS notifications.</p>
+                <Form.Group controlId='phone'>
+                    <Form.Label>Phone Number</Form.Label>
+                    {phoneInput}
+                    {notifyMeAttendeeInput}
+                    {notifyMeHostInput}
+                </Form.Group>
+                <Button variant="primary" type="submit" disabled={props.disabled}>Save</Button>
+            </Form>
+        </div>
+    );
+}
+
+export function PreferencesPage(props: PageProps) {
+    if (!props.user) {
+        redirectToLogin(props.loginUrl);
+    }
+    
+    if (!props.user) throw new Error("user is undefined!");
+    const userId = props.user.id
+
+    // Setup basic state
+    const [user, setUser] = useState(undefined as MyUser | undefined);
+    const [doRefresh, refreshLoading, refreshError] = usePromise(() => api.getUser(userId) as Promise<MyUser>, setUser);
```

Comment:
Good catch - I added `refreshError` to `FormErrors`, but (on other pages at least) `isChanging` is for when you've submitted a change and are waiting for it to be acknowledged. And on other pages, we don't show the `LoadingDisplay` when we're awaiting initial page load. The original reason for that is because the auto-refresher would also trigger the `LoadingDisplay` and some users found it distracting, but that isn't a problem anymore, so it probably makes sense to show the spinner across the board upon page load now. For this PR though, I'm going to leave it consistent with other pages.

Reply:
Okay, that makes sense.